db/versions/0023.yml

```diff
@@ -0,0 +1,1023 @@
+    serviceName: queue
+    args: task_id text, run_id int, state text, reason text, retry_reason text
+    returns: table(retries_left integer, runs jsonb, taken_until timestamptz)
+    body: |-
+      declare
+        task record;
+        run jsonb;
+        new_runs jsonb;
+        new_taken_until timestamptz;
+      begin
+        -- lock the task row to prevent concurrent updates
+        select tasks.retries_left, tasks.runs
+        into task
+        from tasks
+        where tasks.task_id = resolve_task.task_id
+        for update;
+
+        if task.runs is null then
+          return;
+        end if;
+
+        if jsonb_array_length(task.runs) != run_id + 1 then
+          -- run_id is not the latest run
+          return;
+        end if;
+
+        run = task.runs -> run_id;
+        if run ->> 'state' != 'running' then
+          -- run is not running
+          return;
+        end if;
+
+
+        -- reconstruct the task.runs object with an updated run
+        new_runs = (task.runs - run_id) || jsonb_build_array(
+          run || jsonb_build_object(
+            'state', state,
+            'reasonResolved', reason,
+            'resolved', now()));
+
+        -- add a retry if there are any left
+        if retry_reason is not null and task.retries_left > 0 then
+          new_runs = new_runs || jsonb_build_array(
+            jsonb_build_object(
+              'state', 'pending',
+              'reasonCreated', retry_reason,
+              'scheduled', now()));
+          task.retries_left = task.retries_left - 1;
+        end if;
+
+        update tasks
+        set
+          retries_left = task.retries_left,
+          runs = new_runs,
+          taken_until = null
+        where tasks.task_id = resolve_task.task_id;
+
+        return query
+        select tasks.retries_left, tasks.runs, tasks.taken_until
+        from tasks
+        where tasks.task_id = resolve_task.task_id;
+      end
+  check_task_claim:
+    description: |-
+      Check the given task for a claim on the given run expiring at the given
+      time.  If the run is still running, it is marked as claim-expired and
+      a retry scheduled (if retries_left).  
+
+      This returns the task's updated status, or nothing if the current status
+      was not as expected.
+    mode: write
+    serviceName: queue
+    args: task_id text, run_id int, taken_until_in timestamptz
+    returns: table(retries_left integer, runs jsonb, taken_until timestamptz)
+    body: |-
+      declare
+        task record;
+        runs jsonb;
+        run jsonb;
+        new_runs jsonb;
+        new_taken_until timestamptz;
+      begin
+        -- lock the task row to prevent concurrent updates
+        select tasks.retries_left, tasks.runs, tasks.deadline
+        into task
+        from tasks
+        where
+          tasks.task_id = check_task_claim.task_id and
+          tasks.taken_until = taken_until_in
+        for update;
+        raise log 'task %', task;
```

Comment:
do we actually want to be logging this every time? maybe a more descriptive message in that case?

Reply:
yikes, no -- any `raise log` is definitely for debugging only and should be removed